blaze-client/src/main/scala/org/http4s/blaze/client/BlazeClient.scala

```diff
@@ -86,49 +88,53 @@ object BlazeClient {
+            .use { next =>
+              val res: F[Resource[F, Response[F]]] = next.connection
+                .runRequest(req)
+                .adaptError { case EOF =>
+                  new SocketException(s"HTTP connection closed: ${key}")
+                }
+                .map { (response: Resource[F, Response[F]]) =>
+                  response
+                    .flatMap(r => Resource.make(F.pure(r))(_ => manager.release(next.connection)))
+                }
 
-            responseHeaderTimeout match {
-              case responseHeaderTimeout: FiniteDuration =>
-                Deferred[F, Unit].flatMap { gate =>
-                  val responseHeaderTimeoutF: F[TimeoutException] =
-                    F.delay {
-                      val stage =
-                        new ResponseHeaderTimeoutStage[ByteBuffer](
-                          responseHeaderTimeout,
-                          scheduler,
-                          ec,
-                        )
-                      next.connection.spliceBefore(stage)
-                      stage
-                    }.bracket(stage =>
-                      F.asyncF[TimeoutException] { cb =>
-                        F.delay(stage.init(cb)) >> gate.complete(())
-                      }
-                    )(stage => F.delay(stage.removeStage()))
+              responseHeaderTimeout match {
+                case responseHeaderTimeout: FiniteDuration =>
+                  Deferred[F, Unit].flatMap { gate =>
+                    val responseHeaderTimeoutF: F[TimeoutException] =
+                      F.delay {
+                        val stage =
+                          new ResponseHeaderTimeoutStage[ByteBuffer](
+                            responseHeaderTimeout,
+                            scheduler,
+                            ec,
+                          )
+                        next.connection.spliceBefore(stage)
+                        stage
+                      }.bracket(stage =>
+                        F.asyncF[TimeoutException] { cb =>
+                          F.delay(stage.init(cb)) >> gate.complete(())
+                        }
+                      )(stage => F.delay(stage.removeStage()))
 
-                  F.racePair(gate.get *> res, responseHeaderTimeoutF)
-                    .flatMap[Resource[F, Response[F]]] {
-                      case Left((r, fiber)) => fiber.cancel.as(r)
-                      case Right((fiber, t)) => fiber.cancel >> F.raiseError(t)
-                    }
-                }
-              case _ => res
+                    F.racePair(gate.get *> res, responseHeaderTimeoutF)
+                      .flatMap[Resource[F, Response[F]]] {
+                        case Left((r, fiber)) => fiber.cancel.as(r)
+                        case Right((fiber, t)) => fiber.cancel >> F.raiseError(t)
+                      }
+                  }
+                case _ => res
+              }
+            }
+            .recoverWith {
```

Comment:
Concern: are we certain this only borrows one connection at a time?  This is the golden rule of the connection pool, and previous naive flatmaps have deadlocked the connection pool.

Reply:
I gave the client a max of one connection per key, so I think we're good.